src/qml/qmlpreferencesproxy.cpp

```diff
@@ -0,0 +1,262 @@
+#include "qmlpreferencesproxy.h"
+
+#include <qqmlengine.h>
+
+#include <memory>
+
+#include "moc_qmlpreferencesproxy.cpp"
+#include "soundio/soundmanager.h"
+#include "soundio/soundmanagerutil.h"
+#include "util/assert.h"
+#include "util/scopedoverridecursor.h"
+
+namespace mixxx {
+namespace qml {
+
+namespace {
+const QString kAppGroup = QStringLiteral("[App]");
+const ConfigKey kKeylockEngineCfgkey =
+        ConfigKey(kAppGroup, QStringLiteral("keylock_engine"));
+
+} // namespace
+
+QmlSoundDeviceProxy::QmlSoundDeviceProxy(
+        SoundDevicePointer pInternal, Type type, QObject* parent)
+        : QObject(parent),
+          m_pInternal(pInternal),
+          m_type(type) {
+}
+uint QmlSoundDeviceProxy::getChannelCount() const {
+    return m_type == Type::Input ? m_pInternal->getNumInputChannels()
+                                 : m_pInternal->getNumOutputChannels();
+}
+SoundDeviceId QmlSoundDeviceProxy::getDeviceId() const {
+    return m_pInternal->getDeviceId();
+}
+
+QList<QmlSoundDeviceConnection*> QmlSoundDeviceProxy::connections(QmlSoundManagerProxy* manager) {
+    QList<QmlSoundDeviceConnection*> connections;
+
+    auto pManager = manager->internal();
+    auto config = pManager->getConfig();
+
+    if (m_type == Type::Input) {
+        const auto inputDeviceMap = config.getInputs();
+        for (auto it = inputDeviceMap.cbegin(); it != inputDeviceMap.cend(); ++it) {
+            if (it.key() == getDeviceId()) {
+                connections.push_back(new QmlSoundDeviceConnection(
+                        std::make_unique<AudioInput>(it.value()), this));
+            }
+        }
+    } else {
+        const auto ouputDeviceMap = config.getOutputs();
+        for (auto it = ouputDeviceMap.cbegin(); it != ouputDeviceMap.cend(); ++it) {
+            if (it.key() == getDeviceId()) {
+                connections.push_back(new QmlSoundDeviceConnection(
+                        std::make_unique<AudioOutput>(it.value()), this));
+            }
+        }
+    }
+    for (auto* pConnection : connections) {
+        QQmlEngine::setObjectOwnership(pConnection, QQmlEngine::JavaScriptOwnership);
+    }
```

Comment:
this has been done often enough that we should make a wrapper. Something like:

```cpp
template<typename T, typname Args...>
T* make_javascript_owned<T>(args Args&&...) {
    T* p = new T(std::forward<Args...>(args...));
    QQmlEngine::setObjectOwnership(p, QQmlEngine::JavaScriptOwnership);
    return p
}
```
That way, I don't flip out every time I see a naked `new`.  
Ideally, also consider creating a wrapper similar to `parented_ptr`.

Reply:
I went ahead and picked the change here, since it looks like this PR might go ahead before the other one